fix : Fixes Bulk Update failure when getDataValue is involved #11041

Mhashh opened the pull request:
<!--
Thanks for wanting to fix something on Sequelize - we already love you!
Please fill in the template below.
If unsure about something, just do as best as you're able.
-->

## Pull Request Checklist



- [X] Have you added new tests to prevent regressions?
- [ ] If a documentation update is necessary, have you opened a PR to [the documentation repository](https://github.com/sequelize/website/)? <!-- Put PR link here -->
- [ ] Did you update the typescript typings accordingly (if applicable)?
- [X] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [X] Does the name of your PR follow [our conventions](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md#6-commit-your-modifications)?



## Description Of Change
closes [#11041](https://github.com/sequelize/sequelize/issues/11041) in v6 branch.

ephys commented on test/smoketests/smoketest.test.js:
```diff
@@ -0,0 +1,62 @@
+  DataTypes = require('sequelize/lib/data-types'),
+  current = Support.sequelize;
+
+describe(Support.getTestDialectTeaser('Model'), () => {
```
You need to add your test in the integration test folder

Mhashh replied:
okay i will make the changes.

ephys reviewed (changes requested):
Unfortunately we require the fix to land in `main` before we can consider backporting it to v6

Mhashh commented:
> Unfortunately we require the fix to land in `main` before we can consider backporting it to v6

I was thinking since virtual getter is deprecated in v7 and this bug is related to virtual getter , should we add this fix in main?

Mhashh commented:
i have made a pull request for main branch here [https://github.com/sequelize/sequelize/pull/15741.](main pull request)

ephys reviewed (changes requested):
Same issues as reported on https://github.com/sequelize/sequelize/pull/15741

WikiRik commented on test/integration/smoketest.test.js:
```diff
@@ -0,0 +1,64 @@
+  current = Support.sequelize,
+  Op = Sequelize.Op;
+
+describe(Support.getTestDialectTeaser('Model'), () => {
```
Move this test to https://github.com/Mhashh/mhashh-sequelize-fork/blob/v6/test/integration/model.test.js

Mhashh replied:
completed

Mhashh commented:
> Same issues as reported on #15741

I have added separate check for **customgetters** in **get** method.
There are two types of model objects that calls **get** method

1. it has this._options.attributes initialized plus other informational _option properties.
2. it has this._options.attributes set to undefined and no properties of _option set.( through update() , toJSON() etc. method)

So currently get() checks custom getter using
`this._options.attributes  && !this._options.attributes.includes(_key)`

It simply will call all the getters (field with getters, virtual field with getters and purely getter). So in the **issue** a **field with getter** was called and the model object was of second type.


update() builds a temp model object of second type with  **values to be updated** only but it will also try to get all the custom getters value. So when a **field with custom getter** that is **not part of update fields**.
To handle above issue

- if we add an undefined check then it will break calls for getting purely getter methods.
- If we add then check for purely getter method  then it will break calls for virtual fields with getters.
- for checking field type, need access to model class properties like raw attributes.

So I have added another check when this._options.attributes is undefined.

Mhashh commented on src/model.js:
```diff
@@ -3670,19 +3670,24 @@ class Model {
+        for (_key in this._customGetters) {          
+          if (!this._options.attributes.includes(_key)) {
             continue;
           }
-
-          if (Object.prototype.hasOwnProperty.call(this._customGetters, _key)) {
-            values[_key] = this.get(_key, options);
-          }
+          values[_key] = this.get(_key, options);          
+        }
+      }
+      //called by .get() in methods like toJSON, update etc. where this._options is not initialized
+      //so getters in custom getters can have missing datavalue[key](field + getter) or can be purely getter method
+      //above two cases cant be handled simultaneously by above logic
+      else if (this._hasCustomGetters) {
+        const notPureGetter = Object.keys(this.constructor.rawAttributes);
+        for (_key in this._customGetters) {          
+          if (!Object.prototype.hasOwnProperty.call(this.dataValues, _key) && notPureGetter.includes(_key)) {
+            continue;
+          }          
+          values[_key] = this.get(_key, options);          
```
good suggestion, will do the changes.
